compiler/rustc_driver_impl/src/args.rs

```diff
@@ -1,49 +1,73 @@
-use std::error;
-use std::fmt;
-use std::fs;
-use std::io;
+use rustc_session::config::ErrorOutputType;
+use rustc_span::ErrorGuaranteed;
+use std::{env, error, fmt, fs, io};
 
-fn arg_expand(arg: String) -> Result<Vec<String>, Error> {
+fn arg_expand(arg: &str) -> Result<Vec<String>, Error> {
     if let Some(path) = arg.strip_prefix('@') {
         let file = match fs::read_to_string(path) {
             Ok(file) => file,
             Err(ref err) if err.kind() == io::ErrorKind::InvalidData => {
-                return Err(Error::Utf8Error(Some(path.to_string())));
+                return Err(Error::Utf8Error(path.to_string()));
             }
             Err(err) => return Err(Error::IOError(path.to_string(), err)),
         };
         Ok(file.lines().map(ToString::to_string).collect())
     } else {
-        Ok(vec![arg])
+        Ok(vec![arg.to_string()])
     }
 }
 
-pub fn arg_expand_all(at_args: &[String]) -> Vec<String> {
-    let mut args = Vec::new();
+pub fn arg_expand_all(at_args: &[String]) -> Result<Vec<String>, ErrorGuaranteed> {
```

Comment:
I know this is pre-existing, but since you probably spent the time to figure out what this function does and what it is for, please add a doc comment explaining that.